keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AuthorizationSupport.kt

```diff
@@ -18,59 +18,171 @@
+ * optionally, for cases where the API will trigger actuation using the service account associated with the delivery
+ * config, that the caller has access to that service account. Finally, if the API will attempt to read information
+ * from the cloud provider directly (e.g. when retrieving resources to display on the UI), access to the cloud
+ * account may also be checked for each applicable resource (implementors of [Locatable]).
+ *
+ * Use the following guidelines to determine what level of check is required for each API:
+ *   - Operations that read data from keel’s database exclusively should require READ access to the application.
+ *   - Operations that read data from keel’s database and from cloud infrastructure should require READ access to
+ *     the application AND, for each resource returned, READ access to the corresponding cloud account.
+ *   - Operations that trigger storing data in keel’s database, but do NOT trigger infrastructure changes, should
+ *     require WRITE access to the application.
+ *   - Operations that trigger actuation, whereby the service account will take action on behalf of the original caller,
+ *     should check the caller has access to the service account, in addition to WRITE access to the application.
+ */
 @Component
 class AuthorizationSupport(
   private val permissionEvaluator: FiatPermissionEvaluator,
-  private val repository: KeelRepository
+  private val repository: KeelRepository,
+  private val dynamicConfigService: DynamicConfigService
 ) {
   val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
 
-  fun userCanModifyResource(name: String): Boolean =
-    try {
-      val resource = repository.getResource(name)
-      userCanModifySpec(resource.serviceAccount, resource.id)
-    } catch (e: NoSuchResourceException) {
-      // If resource doesn't exist return true so a 404 is propagated from the controller.
-      true
-    }
+  enum class Action {
+    READ, WRITE;
+    override fun toString() = name.toLowerCase()
+  }
 
-  fun userCanModifySpec(serviceAccount: String, specOrName: Any): Boolean {
-    val auth = SecurityContextHolder.getContext().authentication
-    return userCanAccessServiceAccount(auth, serviceAccount, "Resource: $specOrName")
+  enum class TargetEntity {
+    APPLICATION, DELIVERY_CONFIG, RESOURCE;
+    override fun toString() = name.toLowerCase()
   }
 
-  fun userCanModifyApplication(application: String): Boolean = try {
-    val deliveryConfig = repository.getDeliveryConfigForApplication(application)
-    userCanModifyDeliveryConfig(deliveryConfig.serviceAccount, application, deliveryConfig.name)
-  } catch (e: NoSuchElementException) {
-    // If there are no delivery configs for that application return true so this is handled correctly in the controller.
-    true
+  private fun enabled() = dynamicConfigService.isEnabled("keel.authorization", true)
```

Comment:
just out of curiosity, why didn't you use `@ConditionalOnProperty("xxx.enabled")` as in other classes?

Reply:
If I used `@ConditionalOnProperty`, then the entire bean would be missing, and the `@PreAuthorize` annotations would break, so I opted for checking inside the methods.

Reply:
Ah, also, more importantly, this is using a Fast Property, not Spring config! 😃